lua_modules/liquidcrystal/liquidcrystal.lua

```diff
@@ -0,0 +1,213 @@
+local LiquidCrystal = {}
+
+-- commands
+local LCD_CLEARDISPLAY = 0x01
+local LCD_RETURNHOME = 0x02
+local LCD_ENTRYMODESET = 0x04
+local LCD_DISPLAYCONTROL = 0x08
+local LCD_CURSORSHIFT = 0x10
+local LCD_FUNCTIONSET = 0x20
+local LCD_SETCGRAMADDR = 0x40
+local LCD_SETDDRAMADDR = 0x80
+
+-- flags for display entry mode
+-- local LCD_ENTRYRIGHT = 0x00
+local LCD_ENTRYLEFT = 0x02
+local LCD_ENTRYSHIFTINCREMENT = 0x01
+-- local LCD_ENTRYSHIFTDECREMENT = 0x00
+
+-- flags for display on/off control
+local LCD_DISPLAYON = 0x04
+-- local LCD_DISPLAYOFF = 0x00
+local LCD_CURSORON = 0x02
+-- local LCD_CURSOROFF = 0x00
+local LCD_BLINKON = 0x01
+-- local LCD_BLINKOFF = 0x00
+
+-- flags for display/cursor shift
+local LCD_DISPLAYMOVE = 0x08
+local LCD_CURSORMOVE = 0x00
+local LCD_MOVERIGHT = 0x04
+local LCD_MOVELEFT = 0x00
+
+-- flags for function set
+local LCD_8BITMODE = 0x10
+local LCD_4BITMODE = 0x00
+local LCD_2LINE = 0x08
+local LCD_1LINE = 0x00
+local LCD_5x10DOTS = 0x04
+local LCD_5x8DOTS = 0x00
+
+-- defaults
+LiquidCrystal._displayfunction = 0
+LiquidCrystal._displaycontrol = 0
+LiquidCrystal._displaymode = 0
+LiquidCrystal._backlight = false
+
+LiquidCrystal.init = function(self, bus, bus_args, fourbitmode, onelinemode, eightdotsmode, column_width, insert_delay)
+   self._bus_args = bus_args
+   if bus == "I2C" then
+      assert(fourbitmode, "8bit mode not supported")
+      self._backend = dofile "i2c4bit.lua"
```

Comment:
Creating backends like this is going to result in a new object per LCD, even if they all use the same backend.  That's fine, and could even be what you want, except that the backends here are coded to be immutable collections of functions.  So I'd suggest either:

* switch to `require`-ing the backends, so that they only get loaded once (and are subsequently cached)
* redesign the backends to be factories that produce functions that close over their configuration parameters and state, rather than look them up in dictionaries.  These factories can return (or directly set, if that's more your thing) functions that end up being `self._command`, `self._backlight`, `self._write` and friends.

Either approach should help with memory usage.  I think doing both will have the maximum impact, as then the function bodies will be shared and only the closures will vary per LCD.

Reply:
Current code works with both SPIFFS and LFS. `require` will not find module if it is stored on LFS

Reply:
We really should just make `package.loaders` search LFS by default at this point, but thus far the guidance has been for programmers to enable that themselves: https://nodemcu.readthedocs.io/en/master/lfs/#accessing-lfs-functions-and-loading-lfs-modules

Anyway, `require` can be taught to search LFS and it's the recommended configuration.

Failing that, you could push the onus of finding and constructing the backend to the user.  So rather than `:init(bus, bus_args, ...)`, it'd be more like `:init((require "backend")(args), ...)`.